### ISSUE TITLE
New rule: newline-before-if

### ISSUE THREAD

**Issue by fiolkaf** [eslint/eslint#5982]
1. This rule will warn when no newline before if statement.

Correct:
```js
const a = 10;

if (a === 10) {
}
```

Incorrect:
```js
const a = 10;
if (a === 10) {
}
```

2. This rule is stylistic only.

3. This rule is can be widely used as it increases code readability.

4. Yes. I can create rule by myself.

**Comment by platinumazure** [comment]
If we don't already have something like this, I think we should generalize beyond `if`. I think it would probably be better to have a more generic rule, something like `newline-before-control-statement` (but *please* let's find a better name) which has options to turn it on or off for IfStatement, ForStatement, ForInStatement, ForOfStatement, WhileStatement, DoWhileStatement, and whatever constructs I'm not thinking of at the moment.

**Comment by nzakas** [comment]
JSCS has `requireNewlineBeforeBlockStatements`. I'm guessing we'd need the equivalent rather than one that is just specific to `if`.

**Comment by sindresorhus** [comment]
Would be useful to be able to enforce newline after a block too. Like [`requirePaddingNewLinesAfterBlocks`](http://jscs.info/rule/requirePaddingNewLinesAfterBlocks.html). Maybe `space-around-block`.

Similar to http://eslint.org/docs/rules/padded-blocks, but outside blocks.

Related issue: https://github.com/eslint/eslint/issues/3092

**Comment by kaicataldo** [comment]
I can take this on once we've hammered out the details.

**Comment by mysticatea** [comment]
Actually, `requireNewlineBeforeBlockStatements` is covered by our `brace-style` rule with `allman` option.
`requirePaddingNewLinesAfterBlocks` will be covered partly by #5949. But `requirePaddingNewLinesAfterBlocks` is checking more places: in object literal, in call expressions, in new expressions, and in blocks.

`lines-around-block` sounds good to me.
It's similar to `lines-around-comment`. But ignore at the head/tail of a block to avoid conflicting with `padded-blocks` rule.

```js
{
    "lines-around-block": [
        "error",
        {
            "before": "always" | "never" | "any",
            "after": "always" | "never" | "any"
        }
    ]
}
```

```js
/*eslint lines-around-block: ["error", {"before": "always", "after": "always"}]*/

// Bad -------------------------------------------------------------------------
let foo = 0;
if (foo) {    /*error Expected a blank line before this block.*/
    if (foo) {    // Allow nested block. This warned by padded-blocks.
    }
} else {    // Allow 'else' block.
}

let foo = 0;
// a comment
if (foo) {        /*error Expected a blank line before this block.*/
}

// Good ------------------------------------------------------------------------
let foo = 0;

if (foo) {
    if (foo) {
    }
} else {
}

let foo = 0;

// a comment
if (foo) {
}
```

```js
/*eslint lines-around-block: ["error", {"before": "never", "after": "never"}]*/

// Bad -------------------------------------------------------------------------
let foo = 0;
if (foo) {    /*error Expected a blank line before this block.*/
    
    if (foo) {    // Allow nested block. This warned by padded-blocks.
    }
    
} else {    // Allow 'else' block.
}

let foo = 0;

// a comment
if (foo) {        /*error Expected a blank line before this block.*/
}

// Good ------------------------------------------------------------------------
let foo = 0;
if (foo) {
    if (foo) {
    }
} else {
}

let foo = 0;
// a comment
if (foo) {
}
```

```js
/*eslint lines-around-block: ["error", {"before": "any", "after": "any"}]*/

// Bad -------------------------------------------------------------------------

// nothing.

// Good ------------------------------------------------------------------------
let foo = 0;
if (foo) {
    if (foo) {
    }
} else {
}

let foo = 0;

if (foo) {

    if (foo) {
    }

} else {
}
```

We needs to think conflicting with `newline-after-var` and `newline-before-return`.

**Comment by ilyavolodin** [comment]
If we make it into `lines-around-block` we might run into the same issue that we have with `brace-style`, where object declarations are not considered blocks, but people want to be able to control their style in a similar fashion. What if instead we would make a generic rule about line padding, and allow users to specify node types that they want the rule to apply to? Default configuration would be something like:
```js
{ "BlockStatement" : { "before": "always", "after": "never" } }
```

**Comment by mysticatea** [comment]
I think addressing the location of nodes is important for this rule.
Especially, `FunctionExpression` can locate at many places: object methods, class methods, and etc...

In my image, this rule will address bare block statements, block statements with keywords (e.g. if-block, while-block, function-declaration, class-declaration), switch statements.
So...., my proposal's `lines-around-block` is a mix of a part of [requirePaddingNewlinesBeforeKeywords] and a part of [requirePaddingNewLinesAfterBlocks].
In my image, [requirePaddingNewlinesBeforeKeywords] and [requirePaddingNewLinesAfterBlocks] are converted to `lines-around-block` and `lines-between-class-members` or something like (#5949).
But, this customizing of the boundary of rules may cause confusing for JSCS compatibility.
`newline-before-keyword` and `newline-after-block` may be better than my proposal.

[requirePaddingNewlinesBeforeKeywords]: http://jscs.info/rule/requirePaddingNewlinesBeforeKeywords
[requirePaddingNewLinesAfterBlocks]: http://jscs.info/rule/requirePaddingNewLinesAfterBlocks

**Comment by w0rp** [comment]
I'd like to :+1: this one. eslint does a pretty amazing job at enforcing consistent code style. Enforcing blank lines before control statements would be the final piece of the puzzle for me.

**Comment by DinkDonk** [comment]
+1 Looking forward to this addition!

**Comment by TheSavior** [comment]
What is the status of this issue? 

I'm happy to help contribute to this to push it over the finish line. I wrote the JSCS rule this is discussing and this is one of the last rules left in my company's conversion from JSCS to eslint.

**Comment by JonathanWolfe** [comment]
@TheSavior Seems like this feature has stalled out. If you want to implement this and bring it in, I (and probably a lot of others) would greatly appreciate it.

**Comment by TheSavior** [comment]
I think this ticket should be closed as a duplicate of #3092. It sounds like people are in agreement that we should support more than just if statements and that we should merge space after with space before.

**Comment by kaicataldo** [comment]
@TheSavior agreed - sounds like they're converging into the same rule. Will leave open while we discuss the other issue so that we don't lose track of this, since we need it for JSCS compatibility.

**Comment by nzakas** [comment]
@TheSavior @kaicataldo could one or both of you converge these three issues (#5982, #3092, and #5949) into a single issue that summarizes the discussions to this point? I don't think we're going to get very far if we continue to jump back and forth between these issues.

**Comment by quinn** [comment]
superseded by https://github.com/eslint/eslint/issues/7356
